src/azure-cli/azure/cli/command_modules/appservice/custom.py

```diff
@@ -4197,6 +4527,101 @@ def _set_remote_or_local_git(cmd, webapp, resource_group_name, name, deployment_
+    sa_resource_group = resource_group_name
+    
+    if is_valid_resource_id(deployment_storage_name):
+        sa_resource_group = parse_resource_id(deployment_storage_name)['resource_group']
+        deployment_storage_name = parse_resource_id(deployment_storage_name)['name']
+
+    storage_client = get_mgmt_service_client(cli_ctx, StorageManagementClient)
+    storage_properties = storage_client.storage_accounts.get_properties(sa_resource_group, deployment_storage_name)
+    
+    endpoints = storage_properties.primary_endpoints
+    sku = storage_properties.sku.name
+    allowed_storage_types = ['Standard_GRS', 'Standard_RAGRS', 'Standard_LRS', 'Standard_ZRS', 'Premium_LRS', 'Standard_GZRS']
+    
+    if not getattr(endpoints, 'blob', None):
+        raise CLIError("Deployment storage account '{}' has no 'blob' endpoint. It must have blob endpoints enabled".format(deployment_storage_name))
+    
+    if sku not in allowed_storage_types:
+        raise CLIError("Storage type {} is not allowed".format(sku))
+
+    return storage_properties
+
+
+def _normalize_functionapp_name(functionapp_name):
+    import re, string
+    return re.sub(r"[^a-zA-Z0-9]", '', functionapp_name).lower()
+
+
+def _get_or_create_deployment_storage_container(cmd, resource_group_name, functionapp_name, deployment_storage_name, deployment_storage_container_name):
+    storage_client = get_mgmt_service_client(cmd.cli_ctx, StorageManagementClient)
+    if deployment_storage_container_name:
+        storage_container = storage_client.blob_containers.get(resource_group_name, deployment_storage_name, deployment_storage_container_name)
+    else:
+        from random import randint
+        deployment_storage_container_name = "released-package{}{:07}".format(_normalize_functionapp_name(functionapp_name)[:30], randint(0, 9999999))
+        logger.warning("Creating deployment storage account container '%s' ...", deployment_storage_container_name)
+        
+        from azure.mgmt.storage.models import BlobContainer
+
+        storage_container = storage_client.blob_containers.create(resource_group_name, deployment_storage_name, deployment_storage_container_name, BlobContainer())
+
+    logger.warning(storage_container)
+    return storage_container
+
+
+def _get_or_create_user_assigned_identity(cmd, resource_group_name, functionapp_name, user_assigned_identity, location):
+    from azure.mgmt.msi import ManagedServiceIdentityClient
+    msi_client = get_mgmt_service_client(cmd.cli_ctx, ManagedServiceIdentityClient)
+    if user_assigned_identity:
+        if is_valid_resource_id(user_assigned_identity):
+            user_assigned_identity = parse_resource_id(user_assigned_identity)['name']
+        identity = msi_client.user_assigned_identities.get(resource_group_name=resource_group_name, resource_name=user_assigned_identity)
+    else:
+        user_assigned_identity_name = _normalize_functionapp_name(functionapp_name)[:24]
```

Comment:
Might need to change this format in case no function app name characters are alphanumeric